native/src/main/java/io/ballerina/stdlib/data/csvdata/csv/CsvParser.java

```diff
@@ -0,0 +1,1075 @@
+                }
+            }
+            return false;
+        }
+
+        private static void initiateNewRowType(StateMachine sm) {
+            sm.currentCsvNode = CsvCreator.initRowValue(sm.expectedArrayElementType);
+        }
+
+        private static void finalizeTheRow(StateMachine sm) {
+            int rootArraySize = sm.rootArrayType.getSize();
+            if (rootArraySize == -1) {
+                sm.rootCsvNode.append(sm.currentCsvNode);
+            } else if (sm.arraySize < rootArraySize) {
+                sm.rootCsvNode.add(sm.arraySize, sm.currentCsvNode);
+            }
+            sm.arraySize++;
+        }
+
+        private static void addRowValue(StateMachine sm) throws CsvParserException {
+            addRowValue(sm, true);
+        }
+
+        private static void addRowValue(StateMachine sm, boolean trim) throws CsvParserException {
+            Type type;
+            Field currentField = null;
+            sm.isValueStart = false;
+            Type exptype = sm.expectedArrayElementType;
+            String value = sm.value();
+            if (trim) {
+                value = value.trim();
+            }
+
+            if (exptype instanceof RecordType) {
+                type = getExpectedRowTypeOfRecord(sm);
+                currentField = getCurrentField(sm);
+            } else if (exptype instanceof MapType) {
+                type = ((MapType) exptype).getConstrainedType();
+            } else if (exptype instanceof ArrayType) {
+                type = getExpectedRowTypeOfArray(sm, (ArrayType) exptype);
+            } else if (exptype instanceof TupleType) {
+                type = getExpectedRowTypeOfTuple(sm, (TupleType) exptype);
+            } else {
+                throw new CsvParserException("Unexpected expected type");
+            }
+
+            if (type != null) {
+                CsvCreator.convertAndUpdateCurrentJsonNode(sm,
+                        value, type, sm.config, exptype, currentField);
+            }
+            sm.columnIndex++;
+        }
+
+        private static Type getExpectedRowTypeOfTuple(StateMachine sm, TupleType tupleType) {
+            List<Type> tupleTypes = tupleType.getTupleTypes();
+            if (tupleTypes.size() > sm.columnIndex) {
+                return tupleTypes.get(sm.columnIndex);
+            } else {
+                Type restType = sm.restType;
+                if (restType != null) {
+                    return restType;
+                } else {
+                    sm.charBuffIndex = 0;
+                    if (sm.config.allowDataProjection) {
+                        return null;
+                    }
+                    throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_EXPECTED_TUPLE_SIZE, tupleTypes.size());
+                }
+            }
+        }
+
+        private static Type getExpectedRowTypeOfArray(StateMachine sm, ArrayType arrayType) {
+            if (arrayType.getSize() != -1 && arrayType.getSize() <= sm.columnIndex) {
```

Comment:
Could you double check this is working as expected for arrays with inferred sizes?